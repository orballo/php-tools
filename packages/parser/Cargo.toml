[package]
name = "parser"
version.workspace = true
authors.workspace = true
edition.workspace = true
license.workspace = true

[dependencies]
rowan.workspace = true
serde.workspace = true
serde_json.workspace = true
nom.workspace = true
lexer.path = "../lexer"
lexer_derive.path = "../lexer_derive"

[dev-dependencies]
insta.workspace = true

[profile.dev.package]
insta.opt-level = 3
similar.opt-level = 3
