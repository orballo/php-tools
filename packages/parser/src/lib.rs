pub mod php_parser;

pub use php_parser::{PHPParser, Rule as PHPRule};
