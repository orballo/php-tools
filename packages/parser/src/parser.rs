// @TODO: the parser core implementation should parse
// the pieces of the scripts, and once the script sections
// are identified, these are the ones that should be parsed
// with the extensions
//
// The text sections could potentially accept extensions to parse HTML, etc.

pub struct Parser {}

impl Parser {
    pub fn builder() -> ParserBuilder {
        ParserBuilder::new()
    }
}

// @TODO: We need to implement a Token struct.
// This struct is going to be used to store token kinds
// with their name, span, and &str to source.
//
// We need to do the lexing first, and if possible
// concurrently over multiple files.
//
// Parsing should be done after lexing, and only using
// the tokens from the lexer's output.
//
// Each feature/token of the language will be
// in a package called extension.
//
// The individual extensions can implement:
// - Token definition and parser (Optional)
// - Expresion definition and parser (Optional)
// - Statement definition and parser (Optional)
// - Bytecode instruction definition (Optional)
// - Bytecode instruction parser (Optional)
//
pub struct ParserBuilder {
    tokens: Vec<Token>,
    parsers: Vec<Box<dyn SubParser>>,
}

impl ParserBuilder {
    pub fn new() -> Self {
        Self {
            tokens: Vec::new(),
            parsers: Vec::new(),
        }
    }

    pub fn parser() -> Self {
        todo!()
    }

    pub fn parsers() -> Self {
        todo!()
    }
}

pub trait SubParser {}
