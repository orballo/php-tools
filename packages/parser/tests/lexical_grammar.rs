use insta::{assert_debug_snapshot, glob};
use parser::{PHPParser, PHPRule};
use pest::Parser;
use std::fs;

#[test]
fn lexical_grammar_snapshots() {
    let mut settings = insta::Settings::clone_current();
    settings.set_prepend_module_to_snapshot(false);

    settings.bind(|| {
        glob!("fixtures/*.php", |path| {
            let file = fs::read_to_string(path).unwrap();
            let mut parsed_file = PHPParser::parse(PHPRule::SCRIPT, &file)
                .expect("Something went wrong parsing the file");

            assert_debug_snapshot!(parsed_file.next().unwrap());
        })
    });
}
