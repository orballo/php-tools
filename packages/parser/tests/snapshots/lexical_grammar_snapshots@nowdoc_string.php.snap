---
source: packages/parser/tests/lexical_grammar.rs
expression: parsed_file.next().unwrap()
input_file: packages/parser/tests/fixtures/nowdoc_string.php
---
Pair {
    rule: SCRIPT,
    span: Span {
        str: "<?php\n\n<<<'EOD'\nThis is a basic nowdoc string.\nIt can span multiple lines.\nEOD;\n\n?>",
        start: 0,
        end: 83,
    },
    inner: [
        Pair {
            rule: SCRIPT_SECTION,
            span: Span {
                str: "<?php\n\n<<<'EOD'\nThis is a basic nowdoc string.\nIt can span multiple lines.\nEOD;\n\n?>",
                start: 0,
                end: 83,
            },
            inner: [
                Pair {
                    rule: OPEN_TAG,
                    span: Span {
                        str: "<?php",
                        start: 0,
                        end: 5,
                    },
                    inner: [],
                },
                Pair {
                    rule: NOWDOC_STRING_LITERAL,
                    span: Span {
                        str: "<<<'EOD'\nThis is a basic nowdoc string.\nIt can span multiple lines.\nEOD;",
                        start: 7,
                        end: 79,
                    },
                    inner: [
                        Pair {
                            rule: NOWDOC_CONTENT,
                            span: Span {
                                str: "This is a basic nowdoc string.\nIt can span multiple lines.\n",
                                start: 16,
                                end: 75,
                            },
                            inner: [],
                        },
                    ],
                },
                Pair {
                    rule: CLOSE_TAG,
                    span: Span {
                        str: "?>",
                        start: 81,
                        end: 83,
                    },
                    inner: [],
                },
            ],
        },
    ],
}
